File: results/dnd-rest-benchmark/lifecycle-runs/20260731T002352Z_pi_kimi-k2p7-code_python-django/dndsite/http.py
"""HTTP request/response utilities shared by the API views."""

import json

from django.http import HttpResponseBadRequest, JsonResponse


def parse_json(request):
    """Parse the request body as JSON, returning ``None`` on failure."""
    try:
        return json.loads(request.body)
    except (json.JSONDecodeError, TypeError):
        return None


def require_method(request, *methods):
    """Return a 400 response if the request method is not allowed.

    The API returns ``HttpResponseBadRequest`` with an empty body for
    disallowed methods, matching the original behavior. A return value of
    ``None`` means the method is acceptable.
    """
    if request.method not in methods:
        return HttpResponseBadRequest()
    return None


def bad_request(message="invalid request"):
    return JsonResponse({"error": message}, status=400)


def unauthorized(message="invalid credentials"):
    return JsonResponse({"error": message}, status=401)


def conflict(message):
    return JsonResponse({"error": message}, status=409)


def not_found(message):
    return JsonResponse({"error": message}, status=404)


def forbidden(message="forbidden"):
    return JsonResponse({"error": message}, status=403)

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T150315Z_pi_kimi-k2p7-code_typescript-nextjs/app/lib/http.ts
import { NextResponse } from "next/server";

const ErrorMessages = {
  invalidJson: "Invalid JSON",
  badRequest: "Bad request",
  notFound: "Not found",
  conflict: "Conflict",
  unauthorized: "Unauthorized",
  forbidden: "Forbidden",
} as const;

export function ok<T>(data: T) {
  return NextResponse.json(data);
}

export function created<T>(data: T) {
  return NextResponse.json(data, { status: 201 });
}

export function badRequest(message: string = ErrorMessages.badRequest) {
  return NextResponse.json({ error: message }, { status: 400 });
}

export function notFound(message: string = ErrorMessages.notFound) {
  return NextResponse.json({ error: message }, { status: 404 });
}

export function conflict(message: string = ErrorMessages.conflict) {
  return NextResponse.json({ error: message }, { status: 409 });
}

export function unauthorized(message: string = ErrorMessages.unauthorized) {
  return NextResponse.json({ error: message }, { status: 401 });
}

export function forbidden(message: string = ErrorMessages.forbidden) {
  return NextResponse.json({ error: message }, { status: 403 });
}

/**
 * Parse and validate the JSON body of a POST request.
 *
 * Returns the parsed object on success, or a 400 response on failure.  This
 * helper preserves the evaluator's expected error shapes:
 *   - malformed JSON      -> { error: "Invalid JSON" }
 *   - non-object payload  -> { error: "Bad request" }
 */
export async function parseJsonBody(
  req: Request
): Promise<{ ok: true; body: Record<string, unknown> } | { ok: false; response: NextResponse }> {
  let body: unknown;
  try {
    body = await req.json();
  } catch {
    return { ok: false, response: badRequest(ErrorMessages.invalidJson) };
  }

  if (typeof body !== "object" || body === null) {
    return { ok: false, response: badRequest(ErrorMessages.badRequest) };
  }

  return { ok: true, body: body as Record<string, unknown> };
}
